File: scripts/gather-new-data/02_export_pacfish_csv.R
# Author: Saeesh Mangwani
# Date: 2021-05-20

# Description: Exporting pacfish data to CSV format based on a given data range.

# ==== Loading libraries ====
library(DBI)
library(RPostgres)
library(lubridate)
library(rjson)
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(readr)
library(optparse)
source('scripts/help_funcs.R')

# ==== Initializing option parsing ====
option_list <-  list(
  make_option(c("-s", "--startdate"), type="character", default=(Sys.Date() - 30), 
              help="A year month combination indicating the start date for data to post to depth2water. Defaults to 30 days before today. [Default= %default]", 
              metavar="character"),
  make_option(c("-e", "--enddate"), type="character", default=(Sys.Date()), 
              help="A year month combination indicating the end date for data to post to depth2water. Defaults to the current date [Default= %default]", 
              metavar="character")
)

# Parse any provided options and store them in a list
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

# ==== Paths and global variables ====

# Database credentials
creds <- fromJSON(file = 'options/dbase_credentials.json')

# Filepaths for gathering functions
fpaths <- fromJSON(file='options/filepaths.json')

# Ensuring directory exists for holding posting data data
dir_check_create(fpaths$`daily-post-dir`)
# Creating a sub-directory for EC climate, and deleting any prior contents if
# they exist
out_dir <- file.path(fpaths$`daily-post-dir`, 'pacfish')
if(dir.exists(out_dir)) unlink(out_dir, recursive=T)
dir_check_create(out_dir)

# ==== Opening database connection ====

# Opening database connection
conn <- dbConnect(RPostgres::Postgres(),
                  host = creds$host, dbname = creds$dbname,
                  user = creds$user, password = creds$password)

# ==== Daily data ====

# Data query
query <- format_simple_query('pacfish', 'daily', 
                             date_col = 'Date', 
                             start_date = ymd(opt$startdate),
                             end_date = ymd(opt$enddate))

# Reading data
df <- dbGetQuery(conn, query)

# Splitting table by datatype
dfs <- df %>%
  filter(Parameter != 'Air Temperature') %>% 
  group_by(Parameter) %>%
  group_split() %>% 
  map(., ~{
    param <- unique(.x$Parameter)[1] %>% 
      tolower() %>% 
      str_replace_all(., ' ', '_')
    .x %>% 
      select(-numObservations, -Parameter) %>% 
      setNames(c('station_number', 'station_name', 'datetime', param))
  })

# If there are no dfs, it means no new data are available. Skipping the rest...
# Writing if there are data, otherwise printing a nodata message
if(length(dfs) > 0){
  # Joining to a single dataframe
  daily <- dfs[[1]]
  walk(dfs[2:length(dfs)], ~{
    daily <<- full_join(daily, 
                        .x %>% select(-station_name), 
                        by = c('station_number', 'datetime'))
  })
  
  # Writing to csv
  write_csv(daily, file.path(out_dir, 'pacfish-daily.csv'))
  rm(daily)
  gc()
}else{
  print(
    paste0(
      'No new data available for Pacfish between ',
      as.character(opt$startdate),
      ' and ',
      as.character(opt$enddate),
      '. No CSV exported.'
    )
  )
}

# Closing database connection
dbDisconnect(conn)

File: scripts/gather_new_data/02_export_pacfish_csv.py
# Author: Saeesh Mangwani
# Date: 02/08/2023

# Description: Exporting EC Climate data to CSV format based on a provided date range

# %% ===== Loading libraries =====
import os
import sys
from pathlib import Path
os.chdir(Path(__file__).parent.parent.parent)
sys.path.append(os.getcwd())
from json import load
from optparse import OptionParser
from datetime import datetime, timedelta
import psycopg2
import pandas as pd

#%% Initializing option parsing
parser = OptionParser()
parser.add_option(
    "-s", "--startdate", 
    dest="startdate",
    default=(datetime.today() - timedelta(days=31)).strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The start date of the date range for which data are being posted. Defaults to 31 days before today")
parser.add_option(
    "-e", "--enddate", 
    dest="enddate",
    default=datetime.today().strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The end date of the date range for which data are being posted. Defaults to today")
(options, args) = parser.parse_args()

# %% ===== Paths and global variables =====

# Client credentials from JSON
creds = load(open('options/dbase_credentials.json',))

# Filepaths
fpaths = load(open('options/filepaths.json', ))

# Ensuring directory exists for holding posting data data
out_dir = fpaths['update-data-dir']
if not os.path.exists(out_dir):
    os.makedirs(out_dir)

# %% ===== Initializing database connection =====

# Database connection
conn = psycopg2.connect(
    host=creds['host'], 
    port=creds['port'],
    database=creds['dbname'],
    user= creds['user'],
    password=creds['password']
)
cursor = conn.cursor()

# %% ===== Initializing update options =====
start_date = options.startdate
end_date = options.enddate
# start_date = (datetime.today() - timedelta(days=331)).strftime("%Y-%m-%dT00:00:00-00:00")
# end_date =datetime.today().strftime("%Y-%m-%dT00:00:00-00:00")

# %% ==== Gathering update data ====

# Query options
schema = 'pacfish'
table = 'daily'
datecol = 'Date'

# Building query
query = 'select * from {0}.{1} where "{2}" >= \'{3}\' and "{2}" <= \'{4}\''.format(
    schema,
    table,
    datecol,
    start_date,
    end_date
)

# Getting data
cursor.execute(query)
col_names = [i[0] for i in cursor.description]
dat = pd.DataFrame(cursor.fetchall(), columns=col_names)

# %% ==== Formatting update data ====

# Removing air temperature (not a useful parameter)
dat = dat.loc[dat.Parameter != 'Air Temperature']

# Helper functions for splitting tables by parameter
def formatName(param): 
    return str.lower(param).replace(' ', '_')
def splitTable(param):
    outdf = dat.loc[dat.Parameter == param]
    outdf.drop(columns=['numObservations', 'Parameter'], inplace=True)
    outdf.columns = ['station_number', 'station_name', 'datetime', formatName(param)]
    return(outdf)
# Splitting tables and storing as a list
tablist = {formatName(param): splitTable(param) for param in dat.Parameter.unique()}

# Merging tables via full join
daily = tablist.pop(list(tablist.keys())[0])
for key in tablist.keys():
    newtab = tablist.get(key)
    daily = daily.merge(newtab, how='outer', on=['station_number', 'station_name', 'datetime'])
    print('Merged data for parameter ' + key)

# %%  ==== Exporting to CSV ====
if daily.shape[0] == 0:
    print("No new data available for Pacfish between {} and {}. No CSV exported".format(start_date, end_date))
else:
    print("Exporting data to CSV")
    daily.to_csv(out_dir + '/pacfish-daily.csv', index=False)

# %%
